jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1114 @@
+     * Set that the copy need to update.
+     * </p>
+     *
+     * <p>
+     * The created Set's hashCode is equal to the backing Set's hashCode. And the created Set is equal
+     * to another set if they are the same Set or the backing Set is equal to the other Set.
+     * </p>
+     *
+     * <p>
+     * The created set will be serializable if the backing set is serializable.
+     * </p>
+     *
+     *
+     * @param <E> the class of the objects in the set
+     *
+     * @author CHEN Kui
+     * @since Feb 23, 2018
+     */
+
+    private class AsSynchronizedIterateOnCopySet<E>
+            implements Set<E>, Serializable
+    {
+        private static final long serialVersionUID = -102323563687847936L;
+
+        // Backing set.
+        private Set<E> set;
+
+        // Backing set's copy which is unmodifiable. It means the copy set need to be update if the set
+        // is null.
+        private transient Set<E> copy;
+
+
+        /**
+         * Constructor for AsSynchronizedIterateOnCopySet.
+         * @param s the backing graph.
+         */
+        private AsSynchronizedIterateOnCopySet(Set<E> s)
+        {
+            set = Objects.requireNonNull(s, "s must not be null");
+            copy = null;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public int size()
+        {
+            synchronized (mutex) {
+                return set.size();
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isEmpty()
+        {
+            synchronized (mutex) {
+                return set.isEmpty();
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean contains(Object o)
+        {
+            synchronized (mutex) {
+                return set.contains(o);
+            }
+        }
+
+        /**
+         * Returns an iterator over the elements in the backing set's copy which is unmodifiable. The
+         * elements are returned in the same order of the backing set.
+         *
+         * @return an iterator over the elements in the backing set's copy which is unmodifiable.
+         */
+        @Override
+        public Iterator<E> iterator()
+        {
+            return getCopy().iterator();
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public Object[] toArray()
+        {
+            synchronized (mutex) {
+                return set.toArray();
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public <T> T[] toArray(T[] a)
+        {
+            synchronized (mutex) {
+                return set.toArray(a);
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean add(E e)
+        {
+            throw new UnsupportedOperationException("the Set is unmodifiable");
```

Comment:
Don't repeat the String content over and over...use a constant like in UnmodifiableGraph